### PR TITLE
Introduce SimpleCov for specs

### PR DESCRIPTION
**Pull request by mvz**
This can help us find dead or untested code.

### CODE REVIEW

**Comment by troessner** [comment]
How do you intend to use it? As casual rake task?

**Comment by mvz** [comment]
SimpleCov always runs unless told not to, so after each `rake spec` there's up-to-date coverage info in `coverage/`.

**Comment by troessner** [comment]
But then how about making this part of our build? Otherwise this will be lost, at least to me...

**Comment by mvz** [comment]
It is already part of the build. Just run `bundle exec rake spec` as usual and the coverage report will appear in `coverage/`.

Or do you mean adding a check to require a certain coverage level?

**Comment by troessner** [comment]
If it cant make the build fail, its not part of our build. So yes, i think we should have a check for it, otherwise it is meaningless - at least to me.

**Comment by mvz** [comment]
I'll see what I can do.

**Comment by mvz** [comment]
@troessner updated so the build fails when the coverage drops below current levels.

File: spec/reek/cli/application_spec.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+require_relative '../../spec_helper'
+require_lib 'reek/cli/application'
+
+RSpec.describe Reek::CLI::Application do
+  describe '#initialize' do
+    it 'works' do
+      Reek::CLI::Application.new []
+    end
```
Comment (troessner) [inline spec/reek/cli/application_spec.rb]: I dont get it. What's the point of this test? Satisfying simplecov? Shouldnt we rather have proper specs?

Comment (mvz) [inline spec/reek/cli/application_spec.rb]: I had a reasoning (i.e., at least get it loaded so its coverage isn't zero), but I realize it's no better than having zero coverage.

I'll see if I can add some proper specs here and in the Task spec as well.

File: spec/spec_helper.rb [hunk 2]
```diff
@@ -1,4 +1,20 @@
...
+
+SimpleCov.start do
+  track_files 'lib/**/*.rb'
+  add_filter 'lib/reek/version.rb' # version.rb is loaded too early to test
+end
+
+if ENV['FAIL_ON_LOW_COVERAGE'] == '1'
+  SimpleCov.at_exit do
+    SimpleCov.result.format!
+    SimpleCov.minimum_coverage 98.71
```
Comment (andyw8) [inline spec/spec_helper.rb]: Seems having these values will lead to a lot of churn of this file.

One alternative approach I like is to specify 100% coverage and use `# :nocov:` markers for the lines to be ignored.

Comment (troessner) [inline spec/spec_helper.rb]: I'd rather not pollute our source code with "nocov" everywhere.
Especially since I am highly sceptical of simplecov anyways (i'd prefer setting up mutant properly, but thats a different story)

Comment (mvz) [inline spec/spec_helper.rb]: > Seems having these values will lead to a lot of churn of this file.

Possibly, but then again I'm not going to be watching this number very closely and update it every time our coverage increases by 0.01% :laughing:.

My main goal with this is to be able to quickly spot parts of the code that are not used anymore. If an external runner were available (e.g., `simplecov rspec spec`), I'd use that instead.

**Comment by troessner** [comment]
@mvz you're still on this regarding proper specs above, right? 

Just asking, not that you're waiting for a merge and I'm waiting for an update ;)

File: spec/spec_helper.rb [hunk 3]
```diff
@@ -1,4 +1,20 @@
 require 'pathname'
+
+require 'simplecov'
+
+SimpleCov.start do
```
Comment (bf4) [inline spec/spec_helper.rb]: Is there an advantage to putting the config and start here, rather than in '.simplecov.'? It gets loades when simplecov and is required, so no churn in spec helpers

I've also found it best to leave off when not running whole suite as it can slow down the prog exit

**Comment by mvz** [comment]
@troessner Yes, although I got sidetracked on the separate runner idea :-).

File: spec/spec_helper.rb (hunk 3 above)
Comment (mvz) [inline spec/spec_helper.rb]: Good points @bf4, I'll look into them.

**Comment by mvz** [comment]
Tests added. I noticd a bug in Task that I'll create a separate ticket for.

File: lib/reek/cli/reek_command.rb [hunk 4]
```diff
@@ -12,8 +12,8 @@ def execute(app)
         options.sources.each do |source|
           reporter.add_examiner Examiner.new(source, smell_names, configuration: app.configuration)
         end
-        reporter.smells? ? app.report_smells : app.report_success
         reporter.show
+        reporter.smells? ? options.failure_exit_code : options.success_exit_code
```
Comment (troessner) [inline lib/reek/cli/reek_command.rb]: While we are at it, let's rename "options" here to "options_interpreter"!
It's super confusing like it is right now and trips me up every time :-(

File: lib/reek/rake/task.rb [hunk 5]
```diff
@@ -92,6 +92,12 @@ def source_files=(files)
         @source_files = FileList[files]
       end
 
+      def command
```
Comment (troessner) [inline lib/reek/rake/task.rb]: Why this change?

File: spec/quality/reek_source_spec.rb [hunk 6]
```diff
@@ -1,4 +1,5 @@
-require_relative '../spec_helper'
```
Comment (troessner) [inline spec/quality/reek_source_spec.rb]: Isn't requiring the spec_helper kind of best practice even if it's technically not necessary?

I mean, we require files in /lib explicitly for the same reason even though quite a few times it wouldnt be necessary...

File: spec/reek/cli/application_spec.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+  describe '#execute' do
+    let(:command) { double 'reek_command' }
+    let(:app) { Reek::CLI::Application.new [] }
+
+    before do
+      allow(Reek::CLI::ReekCommand).to receive(:new).and_return command
+    end
+
+    it 'executes its command passing itself as a parameter' do
```
Comment (troessner) [inline spec/reek/cli/application_spec.rb]: I'd consider this an implementation detail and rather not have this spec (since refactoring already could break tbis test, which it shouldnt).

File: spec/reek/cli/reek_command_spec.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+        allow(option_interpreter).to receive(:sources).and_return []
+        allow(reporter).to receive(:smells?).and_return true
+      end
+
+      it 'returns a failure code' do
+        result = command.execute app
+        expect(result).to eq Reek::CLI::Options::DEFAULT_FAILURE_EXIT_CODE
+      end
+    end
+  end
```
Comment (troessner) [inline spec/reek/cli/reek_command_spec.rb]: Nice specs!

File: spec/reek/rake/task_spec.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+      task.source_files = '*.rb'
+      expect(task.source_files).to eq FileList['*.rb']
+    end
+  end
+
+  describe '#command' do
+    let(:task) { Reek::Rake::Task.new }
+
+    it 'does not include a config file by default' do
+      expect(task.command).not_to include '-c'
```
Comment (troessner) [inline spec/reek/rake/task_spec.rb]: Hmm, so you made this public just for the test, now i see.
I'd then rather use `send` ....

File: tasks/test.rake [hunk 10]
```diff
@@ -17,6 +17,12 @@ namespace 'test' do
     t.cucumber_opts = 'features --format progress --color'
   end
 
+  task 'enable_simplecov' do
+    ENV['SIMPLECOV'] = '1'
+  end
+
+  task 'spec' => 'test:enable_simplecov'
```
Comment (troessner) [inline tasks/test.rake]: If you enable it on every spec run, why not drop that ENV charade completely and.....just require it unconditionally on spec run?

Comment (mvz) [inline tasks/test.rake]: We don't want to enable it when running one spec file.

Comment (mvz) [inline tasks/test.rake]: Another option is probably to add `-rsimplecov` inside the spec task, so `rspec something_spec.rb` won't activate it.

File: spec/reek/rake/task_spec.rb (hunk 9 above)
Comment (mvz) [inline spec/reek/rake/task_spec.rb]: Yes, that's probably the lesser evil :smile:.

File: spec/reek/cli/application_spec.rb (hunk 7 above)
Comment (mvz) [inline spec/reek/cli/application_spec.rb]: I agree.

File: spec/quality/reek_source_spec.rb (hunk 6 above)
Comment (mvz) [inline spec/quality/reek_source_spec.rb]: This spec is a special case and we don't want to enable simplecov for it.

Comment (troessner) [inline spec/quality/reek_source_spec.rb]: Ok, makes sense. But then I'd prefer to have at least a comment on top of this file screaming in my face that this is special and why...

File: tasks/test.rake (hunk 10 above)
Comment (troessner) [inline tasks/test.rake]: >> We don't want to enable it when running one spec file.

Obvious reason which I completely missed :laughing: 

>> Another option is probably to add -rsimplecov inside the spec task, so rspec something_spec.rb won't activate it.

Ah, I think that's a better solution, let's go for that!

**Comment by mvz** [comment]
Comments processed. @troessner please take another look. @chastell any comments on the RuboCop settings?

**Comment by troessner** [comment]
Looks good to me now and can be merged. @chastell ?

File: spec/reek/cli/application_spec.rb [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+      begin
+        Reek::CLI::Silencer.silently do
+          Reek::CLI::Application.new ['--foo']
+        end
+      rescue SystemExit => e
+        expect(e.status).to eq Reek::CLI::Options::DEFAULT_ERROR_EXIT_CODE
+        error_raised = true
+      end
+      expect(error_raised).to be_truthy
+    end
```
Comment (chastell) [inline spec/reek/cli/application_spec.rb]: Hm, could this be written with the exception matcher? Something to the tune of (untested):

```ruby
call = -> { Reek::CLI::Silencer.silently { Reek::CLI::Application.new ['--foo'] } }
expect(call).to raise_error do |error|
 expect(error).to be_a SystemExit
 expect(error.status).to eq Reek::CLI::Options::DEFAULT_ERROR_EXIT_CODE
end
```

File: spec/reek/rake/task_spec.rb [hunk 12]
```diff
@@ -0,0 +1,39 @@
+require_relative '../../spec_helper'
+require_lib 'reek/rake/task'
+
+RSpec.describe Reek::Rake::Task do
+  describe '#initialize' do
+    it 'works' do
+      Reek::Rake::Task.new
+    end
```
Comment (chastell) [inline spec/reek/rake/task_spec.rb]: I think this should either be `it 'can be called without arguments'` (or whatever this test is testing) or dropped altogether. :)

**Comment by chastell** [comment]
I have two minor nitpicks about tests, but LGTM otherwise. Good work! (Apologies for lagging recently, when I’m not sick I’m catching up with the exam session at the uni…)

File: spec/reek/cli/application_spec.rb (hunk 11 above)
Comment (troessner) [inline spec/reek/cli/application_spec.rb]: I think the rspec matcher only catches StandardError and below which wouldn't work for SystemExit.
Hence the way it is done above ;)

Comment (chastell) [inline spec/reek/cli/application_spec.rb]: Hm, I just did a quick test and the below does work:

```ruby
call = -> { exit 10 }
expect(call).to raise_error(SystemExit)
```

Comment (troessner) [inline spec/reek/cli/application_spec.rb]: Ah, interesting. It was a conjecture from my side, i thought thats why @mvz had written it that way :laughing:

**Comment by chastell** [comment]
> @chastell any comments on the RuboCop settings?

Um, I must be missing something. What RuboCop settings?

File: spec/reek/cli/application_spec.rb (hunk 11 above)
Comment (mvz) [inline spec/reek/cli/application_spec.rb]: I was unaware of the extended matcher with blocks. I'll see if I can make that work.

File: spec/reek/rake/task_spec.rb (hunk 12 above)
Comment (mvz) [inline spec/reek/rake/task_spec.rb]: Oh, I thought I removed this one. It should go.

**Comment by mvz** [comment]
> What RuboCop settings?

The ones that are part of a completely different pull request of course :laughing:.

File: spec/reek/cli/application_spec.rb (hunk 11 above)
Comment (chastell) [inline spec/reek/cli/application_spec.rb]: (In case you’re tempted – or, y’know, TOLD BY RUBOCOP – to go with `expect(call).to raise_error do |error|` just remember that `do` binds weaker than `{` and the block needs to bind to `raise_error`.) ;)

File: tasks/test.rake [hunk 13]
```diff
@@ -4,7 +4,7 @@ require 'rspec/core/rake_task'
 namespace 'test' do
   RSpec::Core::RakeTask.new('spec') do |t|
     t.pattern = 'spec/reek/**/*_spec.rb'
-    t.ruby_opts = ['-Ilib -w']
+    t.ruby_opts = ['-rsimplecov -Ilib -w']
```
Comment (bf4) [inline tasks/test.rake]: :100: : to only requiring (aka starting) simplecov when running the rake spec task and it will always be loaded before any app code!

**Comment by mvz** [comment]
I resolved @chastell's nits. LGTM?

**Comment by troessner** [comment]
I was hoping that my #852 would get merged first so that you had to fix the merge conflicts and not me :laughing: 

Merged!

**Comment by mvz** [comment]
:laughing:
